### ISSUE TITLE
Lost password

### ISSUE THREAD

**Issue by patrik-pwned** [noQ-sweden/noq#647]
**Story**
As a user who has forgotten my password, I want to be able to recover my password using my registered email or Unicode username, So that I can regain access to my account securely.

**Acceptance criterias**
- [x] AC1: Forgot Password Page: The user should be able to enter their registered Unicode, username or email on the "Glömt lösenord" page.
- [x] AC2: Recovery Option (Email): If the user submits a valid email, display confirmation and send a recovery link to their email.
- [x] AC3: Invalid Input Handling: If an unregistered email or username, display an error message.
- [x] AC4: Password Reset Process: The user should be able to reset their password by clicking the email recovery link
- [x] AC5: Password Requirements: Same as during registration.
- [x] AC6: Successful Password Reset: Upon successful password update, redirect the user to the login page with a success message.
- [x] AC7: Security Measures: Invalidate any existing sessions after the new password is set to ensure security.

###Definition of Done
OK from wingmate
Documented
All subtasks done or cancelled

**Comment by patrik-pwned** [comment]
@overath2 överkurs?

**Comment by patrik-pwned** [comment]
bättre?

**Comment by patrik-pwned** [comment]
@overath2 jag gick precis in i appen och man kan klicka på 'glömt lösenord'. Är det så att funktionen (och allt ovan) redan är på plats eller?

**Comment by patrik-pwned** [comment]
Jag fortsätter att gräva

**Comment by overath2** [comment]
Demo januari? Ping @patrik-pwned

**Comment by patrik-pwned** [comment]
@jani-at-noq Jag såg en glömt lösenord funktion när man ska logga in. Har vi redan något av ovannämnda AC på plats? Annars vilka är prio tycker du?

**Comment by jani-at-noq** [comment]
Vi har ingenting på plats. Svårt att säga hur viktigt det är. Kanske efter Jul tänker jag.

**Comment by patrik-pwned** [comment]
Hur funkar det idag? Hur kan en brukare exempelvis logga in om den glömt eller minns fel på sitt lösenord?

**Comment by patrik-pwned** [comment]
> Hur funkar det idag? Hur kan en gäst exempelvis logga in om den glömt eller minns fel på sitt lösenord?

ping @jani-at-noq @overath2

**Comment by overath2** [comment]
Nej det finns inget vad jag vet kring detta. Man måste be @jani-at-noq om reset?

**Comment by patrik-pwned** [comment]
Eller kommer vi undan med manual reset av en 'superadmin' under tiden vi har få användare?

**Comment by patrik-pwned** [comment]
For sprint 15 or 16 @overath2 ? What do you think?

**Comment by overath2** [comment]
This is a very simple basic function. A excellet fisrt story for a new noQ dev.

**Comment by patrik-pwned** [comment]
So sprint 15 for Team Jupiter, or?

**Comment by overath2** [comment]
Yes could be

**Comment by overath2** [comment]
Adding to Neptunus as they have few tickets

**Comment by overath2** [comment]
How much is left? Ping @rohan-noQ and @meelistrm

**Comment by meelistrm** [comment]
I’ve finished the frontend part and I’m now waiting for Rohan to complete the backend. Once that’s done, we’re all set. Last time I spoke to him, he expected it to be ready by the middle of the next sprint😃

**Comment by overath2** [comment]
Cool! If plan is exceeding next sprint please ping me. Then it will be moved to team Venus as we have capacity.

**Comment by patrik-pwned** [comment]
Have a look at [ticket 822](https://github.com/noQ-sweden/noq/issues/822) Becoming a user just in case you need more reference :)

**Comment by overath2** [comment]
When done it is important this feature is handed over to team Venus. Ideally team Venus will be testing this before adding to our common platform where it belongs. @Ovura-noq and @AbubakerSalah please collaborate

**Comment by overath2** [comment]
What is the status here @rohan-noQ ?

**Comment by rohan-noQ** [comment]
We are still in process of creating a new noQ email to send reset links to the users. @AbubakerSalah

**Comment by villenoqnu** [comment]
@meelistrm @rohan-noQ @WrenHuan9 - pause on this ticket until the #834 ( https://github.com/noQ-sweden/noq/issues/834 ) is done.

**Comment by patrik-pwned** [comment]
@meelistrm @rohan-noQ reach out to @overath2 for a handover to Team Venus soon :)

**Comment by overath2** [comment]
Ideally handover to @Ovura-noq or/and @jani-at-noq

**Comment by inhyenoq** [comment]
I have set recover-noreply@noq.nu for this task. Two things are required to check to be able to utilize this email.
1. Check the response message in the image below. It should be written in Swedish (at least) as well. I hope a native Swedish speaker composes a message for this, using a professional and kind tone. 
2. Currently, it rejects all incoming messages from external senders (non-NOQ emails). To utilize it with certain applications as requested, it should allow certain domains to send messages, for example, verification emails. So, please specify the domains to be excluded from the rejection rule.

**Comment by WrenHuan9** [comment]
Hi @rohan-noQ, I noticed a few points that might need further adjustment to fully align with the ticket requirements:

1. Username vs Email for Recovery:
Currently, the recovery flow retrieves the user based on the username. However, users may register with a non-email username, and the ticket specifies that recovery should be possible using either the registered email or Unicode username. This might require further consideration or adjustment.

2. Hardcoded Server Address:
The reset password link currently hardcodes the server address like 'localhost' in the codebase. It might be less convenient when deploying to production environment. Maybe it could be beneficial to abstract this value into a configuration file, such as .env.

3. Password Strength Validation:
At the moment, the new password can be reset even if it does not meet the password requirements specified during registration. To enhance security and meet the acceptance criteria, we might need to enforce the same password validation rules during the reset process.

**Comment by overath2** [comment]
Sorry for close/reopen - my bad.

**Comment by overath2** [comment]
Feedback on the demo: Excellent but the red text needs redesign.

**Comment by WrenHuan9** [comment]
Sorry for the misoperation with status.
Hi @rohan-noQ @meelistrm , I noticed a few points that might need further adjustment to fully align with the ticket requirements:

Username vs Email for Recovery:
Currently, the recovery flow retrieves the user based on the username. However, users may register with a non-email username, and the ticket specifies that recovery should be possible using either the registered email or Unicode username. This might require further consideration or adjustment.

Hardcoded Server Address:
The reset password link currently hardcodes the server address like 'localhost' in the codebase. It might be less convenient when deploying to production environment. Maybe it could be beneficial to abstract this value into a configuration file, such as .env.

Password Strength Validation:
At the moment, the new password can be reset even if it does not meet the password requirements specified during registration. To enhance security and meet the acceptance criteria, we might need to enforce the same password validation rules during the reset process.

**Comment by patrik-pwned** [comment]
Hello everyone,

Thank you all involved for your hard work, guidance and clarity.

So here's my position:
- [ ] Username vs Email for Recovery - **agree, fix this before moving to "Done", please.**
- [ ] Hardcoded Server Address - **please, create a new ticket for next sprint Rohan, unless it's an easy/quick fix.**
- [ ] Password Strength Validation - **agree, fix this before moving to "Done", please.**

**Comment by rohan-noQ** [comment]
Here is the demo link: https://noq2.slack.com/archives/C073F453WFP/p1745844573120959

**Comment by rohan-noQ** [comment]
Thank you for the feedback everyone. Here is my reply, please correct me if I am wrong somewhere thank you.

Username vs Email for Recovery: Currently based on ticket #822 (Becoming a user), there is only one registration method which is by using an email id. From my understanding there is no mechanism in place currently that allows a user to register using an Unicode username. Therefore if it is required to reset passwords using an Unicode username, this ticket will become dependent on #822.

Hardcoded Server Address: Absolutely, it's a good catch, I will fix it and commit the new changes.

Password Strength Validation: Similar to the first point, there is no mechanism in ticket #822 (Becoming a user) to enforce password requirements. Currently a user can register using any basic password (non-empty). Therefore this feature is dependent on ticket #822. 

Design issue pointed by Ove: Absolutely, I will fix it and commit the new changes.

**Comment by overath2** [comment]
This ticket is trending ;)

**Comment by rohan-noQ** [comment]
Sounds good Patrik, will do.
